src/test/regress/expected/failure_truncate.out

```diff
@@ -0,0 +1,1270 @@
+ mitmproxy 
+-----------
+ 
+(1 row)
+
+TRUNCATE test_table;
+ERROR:  canceling statement due to user request
+SELECT citus.mitmproxy('conn.allow()');
+ mitmproxy 
+-----------
+ 
+(1 row)
+
+SELECT * FROM unhealthy_shard_count;
+ count 
+-------
+     0
+(1 row)
+
+SELECT count(*) FROM test_table;
+ count 
+-------
+    20
+(1 row)
+
+-- kill as soon as the coordinator sends TRUNCATE TABLE command
+SELECT citus.mitmproxy('conn.onQuery(query="^TRUNCATE TABLE truncate_failure.test_table").kill()');
+ mitmproxy 
+-----------
+ 
+(1 row)
+
+TRUNCATE test_table;
+ERROR:  server closed the connection unexpectedly
+	This probably means the server terminated abnormally
+	before or while processing the request.
+CONTEXT:  while executing command on localhost:57640
+SELECT citus.mitmproxy('conn.allow()');
+ mitmproxy 
+-----------
+ 
+(1 row)
+
+SELECT * FROM unhealthy_shard_count;
+ count 
+-------
+     0
+(1 row)
+
+SELECT count(*) FROM test_table;
+ count 
+-------
+    20
+(1 row)
+
+-- cancel as soon as the coordinator sends TRUNCATE TABLE command
+SELECT citus.mitmproxy('conn.onQuery(query="^TRUNCATE TABLE truncate_failure.test_table").cancel(' ||  pg_backend_pid() || ')');
+ mitmproxy 
+-----------
+ 
+(1 row)
+
+TRUNCATE test_table;
+ERROR:  canceling statement due to user request
+SELECT citus.mitmproxy('conn.allow()');
+ mitmproxy 
+-----------
+ 
+(1 row)
+
+SELECT * FROM unhealthy_shard_count;
+ count 
+-------
+     0
+(1 row)
+
+SELECT count(*) FROM test_table;
+ count 
+-------
+    20
+(1 row)
+
+-- killing on PREPARE should be fine, everything should be rollbacked
+SELECT citus.mitmproxy('conn.onCommandComplete(command="^PREPARE TRANSACTION").kill()');
+ mitmproxy 
+-----------
+ 
+(1 row)
+
+TRUNCATE test_table;
+ERROR:  connection not open
+CONTEXT:  while executing command on localhost:57640
+SELECT citus.mitmproxy('conn.allow()');
+ mitmproxy 
+-----------
+ 
+(1 row)
+
+SELECT * FROM unhealthy_shard_count;
+ count 
+-------
+     0
+(1 row)
+
+-- we should be able to revocer the transaction and
+-- see that the command is rollbacked
+SELECT recover_prepared_transactions();
+ recover_prepared_transactions 
+-------------------------------
+                             2
+(1 row)
+
+SELECT count(*) FROM test_table;
+ count 
+-------
+    20
+(1 row)
+
+-- cancelling on PREPARE should be fine, everything should be rollbacked
+SELECT citus.mitmproxy('conn.onCommandComplete(command="^PREPARE TRANSACTION").cancel(' ||  pg_backend_pid() || ')');
+ mitmproxy 
+-----------
+ 
+(1 row)
+
+TRUNCATE test_table;
+ERROR:  canceling statement due to user request
+SELECT citus.mitmproxy('conn.allow()');
+ mitmproxy 
+-----------
+ 
+(1 row)
+
+SELECT * FROM unhealthy_shard_count;
+ count 
+-------
+     0
+(1 row)
+
+-- we should be able to revocer the transaction and
+-- see that the command is rollbacked
+SELECT recover_prepared_transactions();
+ recover_prepared_transactions 
```

Comment:
Hard question: Can you explain why we've one prepared transaction here and two in the below test?

Reply:
So, we're not sure why there is a pending prepared transaction in the worker.

That is not related to this PR and seems not a critical issue for now. So, @furkansahin will explain this to @marcocitus and open issue if it is a bug.

Reply:
My expectation is that there are 0 prepared transactions after a cancellation.

What I see in the `cancel()` test is that the `ROLLBACK` is sent before the result for the `PREPARE TRANSACTION` is consumed, which causes it to not roll back.

```
LOG:  issuing ROLLBACK PREPARED 'citus_0_23234_85_15'
DETAIL:  on server localhost:57640
WARNING:  connection error: localhost:57640
DETAIL:  another command is already in progress
WARNING:  failed to roll back prepared transaction 'citus_0_23234_85_15'
HINT:  Run "ROLLBACK PREPARED 'citus_0_23234_85_15'" on localhost:57640
```

That's curious since we do call `ForgetResults` before sending `ROLLBACK PREPARED`, but this might be another `ForgetResults` bug.

Reply:
After the cancellation, we get to this line:
https://github.com/citusdata/citus/blob/master/src/backend/distributed/connection/remote_commands.c#L740

which causes ForgetResults to return early:
https://github.com/citusdata/citus/blob/master/src/backend/distributed/connection/remote_commands.c#L91

It kind of makes sense to bail out of `ForgetResults` on cancellation, since a query might be stuck waiting for network I/O and you might want to stop waiting, even if you're already in the abort handler, but if there was just a single cancellation we should try to consume the result of the PREPARED TRANSACTION and send the ROLLBACK PREPARED.

Reply:
btw, this result seems like a race condition that just happens to trigger almost all of the time.

Reply:
I opened and closed an issue, it's actually expected behaviour because there are 2 cancellations, so the second one ends up cancelling the `ROLLBACK PREPARED`:
https://github.com/citusdata/citus/issues/2287

This still seems like a race condition that occassionally could return a different result. Is there a way to ensure there's only one cancellation?